`assertObjectMatch` is not able to test target object's own `__proto__`

**Describe the bug**

As the title says. `assertEquals` does work expectedly but `assertObjectMatch` does not.

**Steps to Reproduce**

1. Create a file named `test.ts` with the content being:
	
	```typescript
	import { assertObjectMatch, assertEquals } from "jsr:@std/assert@1.0.10"

	Deno.test("assertEquals", async t => {
		await t.step("should be able to test target object's own `__proto__`", async t => {
			const objectA = { ["__proto__"]: { polluted: true } }
			const objectB = { ["__proto__"]: { polluted: true } }
			assertEquals(objectA, objectB)
		})
	})

	Deno.test("assertObjectMatch", async t => {
		await t.step("should be able to test target object's own `__proto__`", async t => {
			const objectA = { ["__proto__"]: { polluted: true } }
			const objectB = { ["__proto__"]: { polluted: true } }
			assertObjectMatch(objectA, objectB)
		})
	})
	```
2. Execute the command `deno test --unstable-unsafe-proto test.ts`
3. See error is like this:
	
	```
	assertEquals ...
	  should be able to test target object's own `__proto__` ... ok (0ms)
	assertEquals ... ok (0ms)
	assertObjectMatch ...
	  should be able to test target object's own `__proto__` ... FAILED (1ms)
	assertObjectMatch ... FAILED (due to 1 failed step) (2ms)
	```
	
	And the diff is even unhelpful:
	
	```
	error: AssertionError: Values are not equal.


	    [Diff] Actual / Expected


	    {}

	  throw new AssertionError(message);
	        ^
	```

**Expected behavior**

The test case successfully passes.

For normal users the infamous `__proto__` should be disabled by default, but as a library author I want to test our APIs are immune to the behavior, but currently I can't use `assignObjectMatch` for it due to this bug.

**Environment**

<!-- please complete the following information -->

- OS: macOS Sequoia 15.0
- deno version: deno 2.1.4 (stable, release, aarch64-apple-darwin)
- std version: @std/assert@1.0.10

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -37,7 +37,9 @@ jobs:
           deno-version: ${{ matrix.deno }}
 
       - name: Run tests
-        run: deno task test
+        run: |
+          deno task test
+          deno task test:with-unsafe-proto
 
       - name: Run timezone-dependent tests
         run: |

--- a/assert/object_match.ts
+++ b/assert/object_match.ts
@@ -137,7 +137,11 @@ function filter(a: loose, b: loose): loose {
           continue;
         }
 
-        defineProperty(filtered, key, filterObject(value as loose, subset as loose));
+        defineProperty(
+          filtered,
+          key,
+          filterObject(value as loose, subset as loose),
+        );
         continue;
       }
 

--- a/assert/object_match.ts
+++ b/assert/object_match.ts
@@ -62,7 +62,7 @@ function defineProperty(target: object, key: PropertyKey, value: unknown) {
 }
 
 function filter(a: loose, b: loose): loose {
-  const seen = new WeakMap();
+  const seen = new WeakMap<object, object>();
   return filterObject(a, b);
 
   function filterObject(a: loose, b: loose): loose {

--- a/assert/object_match.ts
+++ b/assert/object_match.ts
@@ -62,7 +62,7 @@ function defineProperty(target: object, key: PropertyKey, value: unknown) {
 }
 
 function filter(a: loose, b: loose): loose {
-  const seen = new WeakMap<object, object>();
+  const seen = new WeakMap<loose | unknown[], loose | unknown[]>();
   return filterObject(a, b);
 
   function filterObject(a: loose, b: loose): loose {

--- a/assert/object_match.ts
+++ b/assert/object_match.ts
@@ -52,15 +52,23 @@ function isObject(val: unknown): boolean {
   return typeof val === "object" && val !== null;
 }
 
+function defineProperty(target: object, key: PropertyKey, value: unknown) {
+  return Object.defineProperty(target, key, {
+    value,
+    configurable: true,
+    enumerable: true,
+    writable: true,
+  });
+}
+
 function filter(a: loose, b: loose): loose {
   const seen = new WeakMap();
   return filterObject(a, b);
 
   function filterObject(a: loose, b: loose): loose {
     // Prevent infinite loop with circular references with same filter
-    if ((seen.has(a)) && (seen.get(a) === b)) {
-      return a;
-    }
+    const memo = seen.get(a);
+    if (memo && (memo === b)) return a;
 
     try {
       seen.set(a, b);
@@ -82,17 +90,14 @@ function filter(a: loose, b: loose): loose {
     if (keysA.length && keysB.length && !entries.length) {
       // If both objects are not empty but don't have the same keys or symbols,
       // returns the entries in object a.
-      for (const key of keysA) {
-        filtered[key] = a[key];
-      }
-
+      for (const key of keysA) defineProperty(filtered, key, a[key]);
       return filtered;
     }
 
     for (const [key, value] of entries) {
       // On regexp references, keep value as it to avoid loosing pattern and flags
       if (value instanceof RegExp) {
-        filtered[key] = value;
+        defineProperty(filtered, key, value);
         continue;
       }
 
@@ -100,7 +105,7 @@ function filter(a: loose, b: loose): loose {
 
       // On array references, build a filtered array and filter nested objects inside
       if (Array.isArray(value) && Array.isArray(subset)) {
-        filtered[key] = filterArray(value, subset);
+        defineProperty(filtered, key, filterArray(value, subset));
         continue;
       }
 
@@ -108,16 +113,19 @@ function filter(a: loose, b: loose): loose {
       if (isObject(value) && isObject(subset)) {
         // When both operands are maps, build a filtered map with common keys and filter nested objects inside
         if ((value instanceof Map) && (subset instanceof Map)) {
-          filtered[key] = new Map(
-            [...value].filter(([k]) => subset.has(k)).map(
-              ([k, v]) => {
-                const v2 = subset.get(k);
-                if (isObject(v) && isObject(v2)) {
-                  return [k, filterObject(v as loose, v2 as loose)];
-                }
-
-                return [k, v];
-              },
+          defineProperty(
+            filtered,
+            key,
+            new Map(
+              [...value].filter(([k]) => subset.has(k)).map(
+                ([k, v]) => {
+                  const v2 = subset.get(k);
+                  if (isObject(v) && isObject(v2)) {
+                    return [k, filterObject(v as loose, v2 as loose)];
+                  }
+                  return [k, v];
+                },
+              ),
             ),
           );
           continue;
@@ -125,15 +133,15 @@ function filter(a: loose, b: loose): loose {
 
         // When both operands are set, build a filtered set with common values
         if ((value instanceof Set) && (subset instanceof Set)) {
-          filtered[key] = value.intersection(subset);
+          defineProperty(filtered, key, value.intersection(subset));
           continue;
         }
 
-        filtered[key] = filterObject(value as loose, subset as loose);
+        defineProperty(filtered, key, filterObject(value as loose, subset as loose));
         continue;
       }
 
-      filtered[key] = value;
+      defineProperty(filtered, key, value);
     }
 
     return filtered;
@@ -141,9 +149,8 @@ function filter(a: loose, b: loose): loose {
 
   function filterArray(a: unknown[], b: unknown[]): unknown[] {
     // Prevent infinite loop with circular references with same filter
-    if (seen.has(a) && (seen.get(a) === b)) {
-      return a;
-    }
+    const memo = seen.get(a);
+    if (memo && (memo === b)) return a;
 
     seen.set(a, b);
 

--- a/assert/object_match_test.ts
+++ b/assert/object_match_test.ts
@@ -420,7 +420,7 @@ Deno.test(
     const objectA = { ["__proto__"]: { polluted: true } };
     const objectB = { ["__proto__"]: { polluted: true } };
     const objectC = { ["__proto__"]: { polluted: false } };
-    assertObjectMatch(objectA, objectB)
+    assertObjectMatch(objectA, objectB);
     assertThrows(
       () => assertObjectMatch(objectA, objectC),
       AssertionError,
@@ -431,5 +431,5 @@ Deno.test(
       },
     }`,
     );
-  }
+  },
 );

--- a/assert/object_match_test.ts
+++ b/assert/object_match_test.ts
@@ -413,3 +413,23 @@ Deno.test("assertObjectMatch() prints inputs correctly", () => {
     }`,
   );
 });
+
+Deno.test(
+  "assertObjectMatch() should be able to test target object's own `__proto__`",
+  () => {
+    const objectA = { ["__proto__"]: { polluted: true } };
+    const objectB = { ["__proto__"]: { polluted: true } };
+    const objectC = { ["__proto__"]: { polluted: false } };
+    assertObjectMatch(objectA, objectB)
+    assertThrows(
+      () => assertObjectMatch(objectA, objectC),
+      AssertionError,
+      `    {
+      ['__proto__']: {
+-       polluted: true,
++       polluted: false,
+      },
+    }`,
+    );
+  }
+);

--- a/deno.json
+++ b/deno.json
@@ -8,7 +8,8 @@
   },
   "importMap": "./import_map.json",
   "tasks": {
-    "test": "deno test --unstable-http --unstable-webgpu --unstable-unsafe-proto --doc --allow-all --parallel --coverage --trace-leaks --clean",
+    "test": "deno test --unstable-http --unstable-webgpu --doc --allow-all --parallel --coverage --trace-leaks --clean",
+    "test:with-unsafe-proto": "deno test --unstable-http --unstable-webgpu --unstable-unsafe-proto --doc --allow-all --parallel --coverage --trace-leaks --clean",
     "test:browser": "git grep --name-only \"This module is browser compatible.\" | grep -v deno.json | grep -v .github/workflows | grep -v _tools | grep -v encoding/README.md | grep -v media_types/vendor/update.ts | xargs deno check --config browser-compat.tsconfig.json",
     "test:node": "(cd _tools/node_test_runner && npm install) && node --import ./_tools/node_test_runner/register_deno_shim.mjs ./_tools/node_test_runner/run_test.mjs",
     "fmt:licence-headers": "deno run --allow-read --allow-write ./_tools/check_licence.ts",

--- a/deno.json
+++ b/deno.json
@@ -8,7 +8,7 @@
   },
   "importMap": "./import_map.json",
   "tasks": {
-    "test": "deno test --unstable-http --unstable-webgpu --doc --allow-all --parallel --coverage --trace-leaks --clean",
+    "test": "deno test --unstable-http --unstable-webgpu --unstable-unsafe-proto --doc --allow-all --parallel --coverage --trace-leaks --clean",
     "test:browser": "git grep --name-only \"This module is browser compatible.\" | grep -v deno.json | grep -v .github/workflows | grep -v _tools | grep -v encoding/README.md | grep -v media_types/vendor/update.ts | xargs deno check --config browser-compat.tsconfig.json",
     "test:node": "(cd _tools/node_test_runner && npm install) && node --import ./_tools/node_test_runner/register_deno_shim.mjs ./_tools/node_test_runner/run_test.mjs",
     "fmt:licence-headers": "deno run --allow-read --allow-write ./_tools/check_licence.ts",

